# Exploratory Data Analysis

## Import relevant libraries

In [1]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import datetime

## Load data

In [2]:
# setup data path
base_path = os.path.join('..', 'datasets')
company_data_path = os.path.join(base_path, 'raw_data', 'companies.csv')
financial_long_path = os.path.join(base_path, 'raw_data', 'financials__long_term.csv')
financial_short_path = os.path.join(base_path, 'raw_data', 'financials__short_term.csv')

In [3]:
# load datasets
company_data = pd.read_csv(company_data_path)
financial_long = pd.read_csv(financial_long_path)
financial_short = pd.read_csv(financial_short_path)

In [4]:
financial_long

company_name  country   assets    price                  sector  \
0           adidas  germany  12417.0  82.9100  consumer discretionary   
1           adidas  germany  12417.0  84.2000  consumer discretionary   
2           adidas  germany  12417.0  78.4900  consumer discretionary   
3           adidas  germany  12417.0  76.7300  consumer discretionary   
4           adidas  germany  12417.0  78.7100  consumer discretionary   
...            ...      ...      ...      ...                     ...   
10855    vallourec   france   6413.0   4.9300             industrials   
10856    vallourec   france   6413.0   5.0600             industrials   
10857    vallourec   france   6413.0   4.1930             industrials   
10858    vallourec   france   6413.0   2.0850             industrials   
10859    vallourec   france   6413.0   1.6255             industrials   

       operating  debt_to_assets  age                      date  
0           6.08           15.08   64  2014-01-01T00:00:00.000Z  
1           6.08           15.08   64  2014-02-01T00:00:00.000Z  
2           6.08           15.08   64  2014-03-01T00:00:00.000Z  
3           6.08           15.08   64  2014-04-01T00:00:00.000Z  
4           6.08           15.08   64  2014-05-01T00:00:00.000Z  
...          ...             ...  ...                       ...  
10855      -7.07           43.62   61  2018-08-01T00:00:00.000Z  
10856      -7.07           43.62   61  2018-09-01T00:00:00.000Z  
10857      -7.07           43.62   61  2018-10-01T00:00:00.000Z  
10858      -7.07           43.62   61  2018-11-01T00:00:00.000Z  
10859      -7.07           43.62   61  2018-12-01T00:00:00.000Z  

[10860 rows x 9 columns]

## Explore data

There are 182 companies located in `france`, `germany`, `united kingdom`

In [5]:
company_data.head()

country                   name
0  france            accorhotels
1  france  adp aerports de paris
2  france             air france
3  france            air liquide
4  france              axa group

Data in `financial_long` are monthly data from Jan 2014 to Dec 2018  
Balanced panel data: 60 rows each, 181 companies ---- 60*181 = 10,860 rows

In [6]:
financial_long.head()

company_name  country   assets  price                  sector  operating  \
0       adidas  germany  12417.0  82.91  consumer discretionary       6.08   
1       adidas  germany  12417.0  84.20  consumer discretionary       6.08   
2       adidas  germany  12417.0  78.49  consumer discretionary       6.08   
3       adidas  germany  12417.0  76.73  consumer discretionary       6.08   
4       adidas  germany  12417.0  78.71  consumer discretionary       6.08   

   debt_to_assets  age                      date  
0           15.08   64  2014-01-01T00:00:00.000Z  
1           15.08   64  2014-02-01T00:00:00.000Z  
2           15.08   64  2014-03-01T00:00:00.000Z  
3           15.08   64  2014-04-01T00:00:00.000Z  
4           15.08   64  2014-05-01T00:00:00.000Z

In [7]:
len(financial_long['date'].unique())

60

In [8]:
len(financial_long['company_name'].unique())

181

Data in `financial_short` are daily data from 4 Jan 2016 to 30 Dec 2016  
This is an unbalanced panel data probably because holidays are different across country :  
There are companies with 253, 255, 256, 257:  
companies with 253 rows miss date = {'2016-05-02,'2016-05-30,'2016-08-29','2016-12-27'} **ALL UK COMPANIES**   
companies with 255 rows miss date = {'2016-05-16, '2016-10-03'} **ALL GE COMPANIES**  
two companies with 256 rows: {'maurel et prom': '2016-12-30', 'mersen': 2016-08-01'} **BOTH FR COMPANIES**  
companies with 257 rows **GE & FR COMPANIES**

In [9]:
financial_short.head()

company_name  country   assets                      date   price  \
0       adidas  germany  15176.0  2016-01-04T00:00:00.000Z  87.730   
1       adidas  germany  15176.0  2016-01-05T00:00:00.000Z  87.212   
2       adidas  germany  15176.0  2016-01-06T00:00:00.000Z  86.070   
3       adidas  germany  15176.0  2016-01-07T00:00:00.000Z  84.100   
4       adidas  germany  15176.0  2016-01-08T00:00:00.000Z  82.510   

                   sector  operating  debt_to_assets  age  
0  consumer discretionary       7.73           10.66   66  
1  consumer discretionary       7.73           10.66   66  
2  consumer discretionary       7.73           10.66   66  
3  consumer discretionary       7.73           10.66   66  
4  consumer discretionary       7.73           10.66   66

In [10]:
company_dict = {}
for company in financial_short.company_name.unique():
    company_dict[company] = len(financial_short[financial_short['company_name'] == company])

In [11]:
pd.Series(company_dict.values()).unique()

array([257, 255, 253, 256])

In [12]:
company_dict

{'adidas': 257,
 'allianz': 257,
 'basf': 255,
 'bmw': 255,
 'continental': 255,
 'daimler': 255,
 'deutsche borse': 255,
 'lufthansa': 255,
 'deutsche post': 255,
 'eon': 257,
 'heidelberg cement': 255,
 'henkel': 255,
 'rwe': 255,
 'siemens': 255,
 'thyssenkrupp': 255,
 'volkswagen ag': 255,
 'vonovia': 255,
 'elringklinger': 255,
 'leoni ag': 255,
 'cropenergies': 255,
 'evonik industries': 255,
 'fuchs petrolub': 255,
 'lanxess': 255,
 'wacker chemie': 255,
 'alba se': 257,
 'brenntag': 255,
 'cewe stiftung': 255,
 'indus holding': 257,
 'kion group': 255,
 'mtu': 255,
 'osram': 255,
 'wacker neuson': 255,
 'fielmann': 255,
 'hornbach holding': 255,
 'hugo boss': 255,
 'puma': 255,
 'takkt': 255,
 'aareal bank': 255,
 'fraport ag': 255,
 'hamburger hafrn logistik': 255,
 'deutsche wohnen': 255,
 'dic assets': 255,
 'commerzbank': 255,
 'hannover rueck': 255,
 'suedzucker ag': 255,
 'symrise ag': 255,
 'aurubis': 255,
 'talanx': 255,
 'jkx oil and gas': 253,
 'ophir energy': 253,
 '

In [13]:
comp_with_253 = [key for key, values in company_dict.items() if values == 253]
financial_short[financial_short['company_name'].isin(comp_with_253)]

company_name         country  assets                      date  \
12250  jkx oil and gas  united kingdom   247.0  2016-01-04T00:00:00.000Z   
12251  jkx oil and gas  united kingdom   247.0  2016-01-05T00:00:00.000Z   
12252  jkx oil and gas  united kingdom   247.0  2016-01-06T00:00:00.000Z   
12253  jkx oil and gas  united kingdom   247.0  2016-01-07T00:00:00.000Z   
12254  jkx oil and gas  united kingdom   247.0  2016-01-08T00:00:00.000Z   
...                ...             ...     ...                       ...   
32991        rpc group  united kingdom  3535.7  2016-12-22T00:00:00.000Z   
32992        rpc group  united kingdom  3535.7  2016-12-23T00:00:00.000Z   
32993        rpc group  united kingdom  3535.7  2016-12-28T00:00:00.000Z   
32994        rpc group  united kingdom  3535.7  2016-12-29T00:00:00.000Z   
32995        rpc group  united kingdom  3535.7  2016-12-30T00:00:00.000Z   

          price                sector  operating  debt_to_assets  age  
12250   0.37021  energy and materials     -46.48            6.79   23  
12251   0.36498  energy and materials     -46.48            6.79   23  
12252   0.36026  energy and materials     -46.48            6.79   23  
12253   0.35588  energy and materials     -46.48            6.79   23  
12254   0.35335  energy and materials     -46.48            6.79   23  
...         ...                   ...        ...             ...  ...  
32991  11.44587  energy and materials       5.84           32.38   25  
32992  11.35659  energy and materials       5.84           32.38   25  
32993  11.79128  energy and materials       5.84           32.38   25  
32994  11.60407  energy and materials       5.84           32.38   25  
32995  11.63369  energy and materials       5.84           32.38   25  

[20746 rows x 9 columns]

In [14]:
comp_with_255 = [key for key, values in company_dict.items() if values == 255]
financial_short[financial_short['company_name'].isin(comp_with_255)]

company_name  country    assets                      date   price  \
514           basf  germany   76496.0  2016-01-04T00:00:00.000Z  67.580   
515           basf  germany   76496.0  2016-01-05T00:00:00.000Z  67.200   
516           basf  germany   76496.0  2016-01-06T00:00:00.000Z  66.080   
517           basf  germany   76496.0  2016-01-07T00:00:00.000Z  64.470   
518           basf  germany   76496.0  2016-01-08T00:00:00.000Z  63.680   
...            ...      ...       ...                       ...     ...   
12245       talanx  germany  156626.0  2016-12-23T00:00:00.000Z  31.965   
12246       talanx  germany  156626.0  2016-12-27T00:00:00.000Z  32.050   
12247       talanx  germany  156626.0  2016-12-28T00:00:00.000Z  31.910   
12248       talanx  germany  156626.0  2016-12-29T00:00:00.000Z  31.900   
12249       talanx  germany  156626.0  2016-12-30T00:00:00.000Z  31.770   

                     sector  operating  debt_to_assets  age  
514    energy and materials      11.17           21.32  151  
515    energy and materials      11.17           21.32  151  
516    energy and materials      11.17           21.32  151  
517    energy and materials      11.17           21.32  151  
518    energy and materials      11.17           21.32  151  
...                     ...        ...             ...  ...  
12245            financials       7.35            2.23   20  
12246            financials       7.35            2.23   20  
12247            financials       7.35            2.23   20  
12248            financials       7.35            2.23   20  
12249            financials       7.35            2.23   20  

[10965 rows x 9 columns]

In [15]:
comp_with_256 = [key for key, values in company_dict.items() if values == 256]
financial_short[financial_short['company_name'].isin(comp_with_256)]

company_name country  assets                      date  price  \
41477  maurel et prom  france  2396.4  2016-01-04T00:00:00.000Z   2.95   
41478  maurel et prom  france  2396.4  2016-01-05T00:00:00.000Z   2.88   
41479  maurel et prom  france  2396.4  2016-01-06T00:00:00.000Z   2.76   
41480  maurel et prom  france  2396.4  2016-01-07T00:00:00.000Z   2.63   
41481  maurel et prom  france  2396.4  2016-01-08T00:00:00.000Z   2.51   
...               ...     ...     ...                       ...    ...   
41984          mersen  france  1001.2  2016-12-22T00:00:00.000Z  19.42   
41985          mersen  france  1001.2  2016-12-23T00:00:00.000Z  19.74   
41986          mersen  france  1001.2  2016-12-27T00:00:00.000Z  20.20   
41987          mersen  france  1001.2  2016-12-28T00:00:00.000Z  20.30   
41988          mersen  france  1001.2  2016-12-29T00:00:00.000Z  20.30   

                     sector  operating  debt_to_assets  age  
41477  energy and materials       5.27           30.50  175  
41478  energy and materials       5.27           30.50  175  
41479  energy and materials       5.27           30.50  175  
41480  energy and materials       5.27           30.50  175  
41481  energy and materials       5.27           30.50  175  
...                     ...        ...             ...  ...  
41984           industrials       4.06           23.77   79  
41985           industrials       4.06           23.77   79  
41986           industrials       4.06           23.77   79  
41987           industrials       4.06           23.77   79  
41988           industrials       4.06           23.77   79  

[512 rows x 9 columns]

In [16]:
comp_with_257 = [key for key, values in company_dict.items() if values == 257]
financial_short[financial_short['company_name'].isin(comp_with_257)].groupby('company_name')['country'].agg('unique')

company_name
accorhotels               [france]
adidas                   [germany]
adp aerports de paris     [france]
air france                [france]
air liquide               [france]
alba se                  [germany]
allianz                  [germany]
bnp paribas               [france]
bonduelle                 [france]
bouygues                  [france]
bureau veritas            [france]
carrefour                 [france]
danone                    [france]
edf                       [france]
engie                     [france]
eon                      [germany]
gecina                    [france]
groupe casino             [france]
groupe eurotunnel         [france]
hermes international      [france]
indus holding            [germany]
ipsos                     [france]
kering                    [france]
klepierre                 [france]
l'oreal                   [france]
legrand                   [france]
lvmh                      [france]
michelin                  [france]
nexans 

Companies that exist in `financial_long` but not in `financial_short`  

In [17]:
set(financial_long.company_name.unique()) - set(financial_short.company_name.unique())

{'axa group',
 'baywa munich',
 'beiersdorf',
 'euromoney insttutional investors plc',
 'faurecia',
 'king fisher',
 'munich re',
 'rekitt benckise',
 'rightmove',
 'schneider electric',
 'sse plc',
 'victrex group'}

Companies that exist in `financial_short` but not in `financial_long`

In [18]:
set(financial_short.company_name.unique()) - set(financial_long.company_name.unique())

{'leoni ag'}

In [19]:
financial_long.dtypes

company_name       object
country            object
assets            float64
price             float64
sector             object
operating         float64
debt_to_assets    float64
age                 int64
date               object
dtype: object

## Transform date column into datetime format

In [20]:
# transform date series into datetime format
date = pd.to_datetime(financial_long['date'])
# set the format to Year-Month-Date
financial_long.loc[:, 'date'] = date.dt.strftime('%Y-%m-%d')

# decompose date column
financial_long.loc[:, 'year'] = date.dt.strftime('%Y')
financial_long.loc[:, 'month'] = date.dt.strftime('%m')
financial_long.loc[:, 'day'] = date.dt.strftime('%d')

In [21]:
# transform date series into datetime format
date = pd.to_datetime(financial_short['date'])
# set the format to Year-Month-Date
financial_short.loc[:, 'date'] = date.dt.strftime('%Y-%m-%d')

# decompose date column
financial_short.loc[:, 'year'] = date.dt.strftime('%Y')
financial_short.loc[:, 'month'] = date.dt.strftime('%m')
financial_short.loc[:, 'day'] = date.dt.strftime('%d')


## Rearrange columns

The columns orders are not the same so we decided to rearrange the columns

In [22]:
financial_short = financial_short[financial_long.columns]

In [23]:
financial_short.head()

company_name  country   assets   price                  sector  operating  \
0       adidas  germany  15176.0  87.730  consumer discretionary       7.73   
1       adidas  germany  15176.0  87.212  consumer discretionary       7.73   
2       adidas  germany  15176.0  86.070  consumer discretionary       7.73   
3       adidas  germany  15176.0  84.100  consumer discretionary       7.73   
4       adidas  germany  15176.0  82.510  consumer discretionary       7.73   

   debt_to_assets  age        date  year month day  
0           10.66   66  2016-01-04  2016    01  04  
1           10.66   66  2016-01-05  2016    01  05  
2           10.66   66  2016-01-06  2016    01  06  
3           10.66   66  2016-01-07  2016    01  07  
4           10.66   66  2016-01-08  2016    01  08

In [24]:
financial_long.head()

company_name  country   assets  price                  sector  operating  \
0       adidas  germany  12417.0  82.91  consumer discretionary       6.08   
1       adidas  germany  12417.0  84.20  consumer discretionary       6.08   
2       adidas  germany  12417.0  78.49  consumer discretionary       6.08   
3       adidas  germany  12417.0  76.73  consumer discretionary       6.08   
4       adidas  germany  12417.0  78.71  consumer discretionary       6.08   

   debt_to_assets  age        date  year month day  
0           15.08   64  2014-01-01  2014    01  01  
1           15.08   64  2014-02-01  2014    02  01  
2           15.08   64  2014-03-01  2014    03  01  
3           15.08   64  2014-04-01  2014    04  01  
4           15.08   64  2014-05-01  2014    05  01

# Data Preparation

## For slope chart

In [25]:
# select only necessary columns
df = financial_long[['country', 'sector', 'price', 'operating', 'debt_to_assets', 'year']]
df

country                  sector    price  operating  debt_to_assets  \
0      germany  consumer discretionary  82.9100       6.08           15.08   
1      germany  consumer discretionary  84.2000       6.08           15.08   
2      germany  consumer discretionary  78.4900       6.08           15.08   
3      germany  consumer discretionary  76.7300       6.08           15.08   
4      germany  consumer discretionary  78.7100       6.08           15.08   
...        ...                     ...      ...        ...             ...   
10855   france             industrials   4.9300      -7.07           43.62   
10856   france             industrials   5.0600      -7.07           43.62   
10857   france             industrials   4.1930      -7.07           43.62   
10858   france             industrials   2.0850      -7.07           43.62   
10859   france             industrials   1.6255      -7.07           43.62   

       year  
0      2014  
1      2014  
2      2014  
3      2014  
4      2014  
...     ...  
10855  2018  
10856  2018  
10857  2018  
10858  2018  
10859  2018  

[10860 rows x 6 columns]

In [26]:
# each column shows the average
group = ['country', 'sector', 'year']
values = ['price', 'operating', 'debt_to_assets']
group_df = df.groupby(group)[values].agg('mean')
group_df

price  operating  \
country        sector                 year                           
france         consumer discretionary 2014   110.697222  15.005000   
                                      2015   132.315417  17.016667   
                                      2016   137.288889  15.935000   
                                      2017   186.688403  17.713333   
                                      2018   229.385694  16.960000   
...                                                 ...        ...   
united kingdom utilities              2014  1054.526639  17.546667   
                                      2015  1113.804292  17.971667   
                                      2016  1157.524458  19.243333   
                                      2017  1099.133958  19.806667   
                                      2018   922.422222  20.205000   

                                            debt_to_assets  
country        sector                 year                  
france         consumer discretionary 2014       19.136667  
                                      2015       18.283333  
                                      2016       16.358333  
                                      2017       17.340000  
                                      2018       18.691667  
...                                                    ...  
united kingdom utilities              2014       46.026667  
                                      2015       47.846667  
                                      2016       46.385000  
                                      2017       47.521667  
                                      2018       46.665000  

[90 rows x 3 columns]

In [27]:
group_df.reset_index(inplace = True)
group_df

country                  sector  year        price  operating  \
0           france  consumer discretionary  2014   110.697222  15.005000   
1           france  consumer discretionary  2015   132.315417  17.016667   
2           france  consumer discretionary  2016   137.288889  15.935000   
3           france  consumer discretionary  2017   186.688403  17.713333   
4           france  consumer discretionary  2018   229.385694  16.960000   
..             ...                     ...   ...          ...        ...   
85  united kingdom               utilities  2014  1054.526639  17.546667   
86  united kingdom               utilities  2015  1113.804292  17.971667   
87  united kingdom               utilities  2016  1157.524458  19.243333   
88  united kingdom               utilities  2017  1099.133958  19.806667   
89  united kingdom               utilities  2018   922.422222  20.205000   

    debt_to_assets  
0        19.136667  
1        18.283333  
2        16.358333  
3        17.340000  
4        18.691667  
..             ...  
85       46.026667  
86       47.846667  
87       46.385000  
88       47.521667  
89       46.665000  

[90 rows x 6 columns]

In [28]:
# get the value of year 2014
group = ['country', 'sector']
values = ['year','price', 'operating', 'debt_to_assets']

temp_df = group_df.groupby(group)[values].agg('first')
temp_df.reset_index(inplace = True)
temp_df.rename(columns = {'price': 'price_base',
                          'operating': 'operating_base',
                          'debt_to_assets': 'debt_to_assets_base'}, inplace = True)
temp_df

country                  sector  year   price_base  operating_base  \
0           france  consumer discretionary  2014   110.697222       15.005000   
1           france        consumer staples  2014    66.772847        9.856667   
2           france    energy and materials  2014    47.830811       15.913333   
3           france              financials  2014    46.329310       22.480000   
4           france             industrials  2014    38.482918        8.153158   
5           france               utilities  2014    18.816750        6.950000   
6          germany  consumer discretionary  2014    82.881000       10.391528   
7          germany        consumer staples  2014    52.980889       11.046667   
8          germany    energy and materials  2014    43.950942        7.909000   
9          germany              financials  2014    52.486000       38.032000   
10         germany             industrials  2014    58.476250        8.018421   
11         germany               utilities  2014    20.708667        4.190000   
12  united kingdom  consumer discretionary  2014  1095.541429       17.097778   
13  united kingdom        consumer staples  2014  1650.890130       13.882222   
14  united kingdom    energy and materials  2014   923.568900        2.361889   
15  united kingdom              financials  2014   596.086424       32.065455   
16  united kingdom             industrials  2014   692.936171        7.542500   
17  united kingdom               utilities  2014  1054.526639       17.546667   

    debt_to_assets_base  
0             19.136667  
1             24.625000  
2             27.556667  
3             27.971667  
4             28.934737  
5             23.573333  
6             13.415000  
7              8.570000  
8             19.813000  
9             25.224000  
10            26.073684  
11            18.580000  
12            23.191852  
13            27.582222  
14            15.940667  
15            10.450000  
16            23.669000  
17            46.026667

In [29]:
# merge the two tables
merged_df = pd.merge(group_df, temp_df, on = ['country', 'sector', 'year'], how = 'left')
# fill na with the base values
merged_df.fillna(method = 'ffill', inplace = True)
merged_df

country                  sector  year        price  operating  \
0           france  consumer discretionary  2014   110.697222  15.005000   
1           france  consumer discretionary  2015   132.315417  17.016667   
2           france  consumer discretionary  2016   137.288889  15.935000   
3           france  consumer discretionary  2017   186.688403  17.713333   
4           france  consumer discretionary  2018   229.385694  16.960000   
..             ...                     ...   ...          ...        ...   
85  united kingdom               utilities  2014  1054.526639  17.546667   
86  united kingdom               utilities  2015  1113.804292  17.971667   
87  united kingdom               utilities  2016  1157.524458  19.243333   
88  united kingdom               utilities  2017  1099.133958  19.806667   
89  united kingdom               utilities  2018   922.422222  20.205000   

    debt_to_assets   price_base  operating_base  debt_to_assets_base  
0        19.136667   110.697222       15.005000            19.136667  
1        18.283333   110.697222       15.005000            19.136667  
2        16.358333   110.697222       15.005000            19.136667  
3        17.340000   110.697222       15.005000            19.136667  
4        18.691667   110.697222       15.005000            19.136667  
..             ...          ...             ...                  ...  
85       46.026667  1054.526639       17.546667            46.026667  
86       47.846667  1054.526639       17.546667            46.026667  
87       46.385000  1054.526639       17.546667            46.026667  
88       47.521667  1054.526639       17.546667            46.026667  
89       46.665000  1054.526639       17.546667            46.026667  

[90 rows x 9 columns]

In [30]:
# index table (using 2014 as the base year)

def index(col1, col2):
    formula = np.divide(col1, col2) * 100
    return formula

# perform operation on a temporary dataframe
temp_df = merged_df
temp_df.loc[:, 'price_index'] = index(temp_df['price'], temp_df['price_base'])
temp_df.loc[:, 'operating_index'] = index(temp_df['operating'], temp_df['operating_base'])
temp_df.loc[:, 'debt_to_assets_index'] = index(temp_df['debt_to_assets'], temp_df['debt_to_assets_base'])

# create index table
columns = ['country', 'sector', 'year', 'price_index', 'operating_index', 'debt_to_assets_index']
index_df = temp_df[columns]
index_df

country                  sector  year  price_index  \
0           france  consumer discretionary  2014   100.000000   
1           france  consumer discretionary  2015   119.529121   
2           france  consumer discretionary  2016   124.021982   
3           france  consumer discretionary  2017   168.647775   
4           france  consumer discretionary  2018   207.219016   
..             ...                     ...   ...          ...   
85  united kingdom               utilities  2014   100.000000   
86  united kingdom               utilities  2015   105.621257   
87  united kingdom               utilities  2016   109.767209   
88  united kingdom               utilities  2017   104.230080   
89  united kingdom               utilities  2018    87.472634   

    operating_index  debt_to_assets_index  
0        100.000000            100.000000  
1        113.406642             95.540847  
2        106.197934             85.481623  
3        118.049539             90.611392  
4        113.028990             97.674621  
..              ...                   ...  
85       100.000000            100.000000  
86       102.422112            103.954229  
87       109.669453            100.778534  
88       112.879939            103.248117  
89       115.150076            101.386877  

[90 rows x 6 columns]

In [31]:
# get percentage change
temp_df.loc[:, 'price_index'] = index(temp_df['price'], temp_df['price_base'])

In [32]:
# get percentage change

def get_pct_change(df, col, group):
    # remove false positive warning
    pd.options.mode.chained_assignment = None
    formula = df.groupby(group)[col].pct_change()
    return formula

group = ['country', 'sector']
index_df.loc[:, 'price_pct_chanege'] = get_pct_change(index_df, 'price_index', group)
index_df.loc[:, 'operating_pct_chanege'] = get_pct_change(index_df, 'operating_index', group)
index_df.loc[:, 'debt_to_assets_pct_chanege'] = get_pct_change(index_df, 'debt_to_assets_index', group)
index_df

country                  sector  year  price_index  \
0           france  consumer discretionary  2014   100.000000   
1           france  consumer discretionary  2015   119.529121   
2           france  consumer discretionary  2016   124.021982   
3           france  consumer discretionary  2017   168.647775   
4           france  consumer discretionary  2018   207.219016   
..             ...                     ...   ...          ...   
85  united kingdom               utilities  2014   100.000000   
86  united kingdom               utilities  2015   105.621257   
87  united kingdom               utilities  2016   109.767209   
88  united kingdom               utilities  2017   104.230080   
89  united kingdom               utilities  2018    87.472634   

    operating_index  debt_to_assets_index  price_pct_chanege  \
0        100.000000            100.000000                NaN   
1        113.406642             95.540847           0.195291   
2        106.197934             85.481623           0.037588   
3        118.049539             90.611392           0.359822   
4        113.028990             97.674621           0.228709   
..              ...                   ...                ...   
85       100.000000            100.000000                NaN   
86       102.422112            103.954229           0.056213   
87       109.669453            100.778534           0.039253   
88       112.879939            103.248117          -0.050444   
89       115.150076            101.386877          -0.160774   

    operating_pct_chanege  debt_to_assets_pct_chanege  
0                     NaN                         NaN  
1                0.134066                   -0.044592  
2               -0.063565                   -0.105287  
3                0.111599                    0.060010  
4               -0.042529                    0.077951  
..                    ...                         ...  
85                    NaN                         NaN  
86               0.024221                    0.039542  
87               0.070760                   -0.030549  
88               0.029274                    0.024505  
89               0.020111                   -0.018027  

[90 rows x 9 columns]

In [33]:
# save table as csv
path = os.path.join(base_path, 'processed_data', 'slope_chart.csv')
index_df.to_csv(path)